File: DesafioConcreteSolutions/app/src/main/java/br/com/desafio/ui/adapter/RepositoryAdapter.java
package br.com.desafio.ui.adapter;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;

import org.androidannotations.annotations.EBean;
import org.androidannotations.annotations.RootContext;

import br.com.desafio.domain.Repositories;
import br.com.desafio.domain.Repository;
import br.com.desafio.ui.adapter.bind.RepositoryItemView;
import br.com.desafio.ui.adapter.bind.RepositoryItemView_;
import br.com.desafio.ui.adapter.common.RecyclerViewAdapterBase;
import br.com.desafio.ui.adapter.common.ViewWrapper;
import lombok.Setter;

@EBean
public class RepositoryAdapter extends RecyclerViewAdapterBase<Repository, RepositoryItemView> {
    @RootContext
    Context context;
    @Setter
    AdapterView.OnItemClickListener onItemClickListener;

    public void setItems(Repositories repositories){
        super.items = repositories.getItems();
    }

    public void addItem(Repository repository, int position){
        super.items.add(repository);
        notifyItemInserted(position);
    }

    @Override
    protected RepositoryItemView onCreateItemView(ViewGroup parent, int viewType) {
        return RepositoryItemView_.build(context);
    }

    @Override
    public void onBindViewHolder(ViewWrapper<RepositoryItemView> viewHolder, final int position) {
        RepositoryItemView view = viewHolder.getView();
        Repository repository = items.get(position);

        view.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View v) {
                onItemClickListener.onItemClick(null, v, position, position);
            }
        });

        view.bind(repository);
    }
}

File: app/src/main/java/com/concrete/andresdavid/desafioandroid/adapters/RepositoryAdapter.kt
package com.concrete.andresdavid.desafioandroid.adapters

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.concrete.andresdavid.desafioandroid.R
import com.concrete.andresdavid.desafioandroid.model.Repository
import com.concrete.andresdavid.desafioandroid.util.CircleTransform
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.repository_list_item.view.*

class RepositoryAdapter(private val repositories: List<Repository>, val context: Context)
    : RecyclerView.Adapter<RepositoryHolder>() {

    override fun getItemCount(): Int {
        return repositories.count()
    }

    // Inflates the item views
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RepositoryHolder {
        return RepositoryHolder(LayoutInflater.from(context).inflate(R.layout.repository_list_item, parent, false))
    }

    // Binds each animal in the ArrayList to a view
    override fun onBindViewHolder(holder: RepositoryHolder, position: Int) {
        val itemRepository = repositories[position]
        holder.bindRepository(itemRepository)
    }
}

class RepositoryHolder(v: View) : RecyclerView.ViewHolder(v), View.OnClickListener {

    private var view: View = v
    private var repository: Repository? = null

    init {
        v.setOnClickListener(this)
    }

    override fun onClick(v: View) {
        Log.d("RecyclerView", "CLICK!")
    }

    companion object {
        private val REPOSITORY_KEY = "REPOSITORY"
    }

    fun bindRepository(repository: Repository) {
        this.repository = repository
        Picasso.get().load(repository.owner?.avatarUrl).transform(CircleTransform()).into(view.image_user)

        //TODO("validate when image does not work or does not exist")

        view.tv_repository_name.text = repository.name
        view.tv_repository_description.text = repository.description
        view.tv_user_name.text = repository.owner?.login
        view.tv_stars_count.text = repository.stargazersCount.toString()
        view.tv_fork_count.text = repository.forksCount.toString()
    }
}